[SRA] Reshaping SRA activity into vignette (#24)
fhdsl/Data_on_AnVIL: 03-importing_from_SRA.Rmd

The SRA Data corresponding to this project is located [here](https://www.ncbi.nlm.nih.gov/Traces/study/?acc=SRP409181&o=acc_s%3Aa).

```{r, fig.align='center', echo = FALSE, fig.alt= "Microbiome diversity has many beneficial properties, ranging from microbial properties to soil and plant health."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1l0P0gFpsPkYG7blqJ_5JyYYlztJFZDD39CnIB4svrY8/edit#slide=id.g208b8f790dc_23_217")
```

::: {.notice}
_Genetics_

**Novice**: no genetics skills needed

_Programming skills_

**Intermediate**: no programming skills needed
:::

::: {.notice}
**What will this cost?**

You might hear new terms for moving data around in the cloud. **Ingress** is when data comes to you, similar to downloading a file or receiving an email with an attachment. **Egress** is sending the data to another resource, similar to uploading or sending an attached file via email. **There is no fee for ingressing data to AnVIL from SRA**, but there is a small cost for storing the data in an AnVIL bucket.

> There is also no cost for importing a workflow into your workspace, but there may be a cost for running a workflow..

Reviewer: ```suggestion
There is also no cost for importing a workflow into your workspace, but there is a small cost for running a workflow..
```

Reviewer: Could use estimate here! :) https://anvil.terra.bio/#workspaces/anvil-outreach/import_data_sra/submission_history/79fa647e-c147-4c40-a391-1cf61036f8d2

Author: Added!

ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3ce9c5d5921_0_36#slide=id.g3ce9c5d5921_0_36")
```

Once you have opened the workspace page, create a new workspace by clicking on the plus sign at the top.

```{r, echo=FALSE, fig.alt="Create a new workspace by clicking on the plus sign at the top."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3ce9c5d5921_0_46#slide=id.g3ce9c5d5921_0_46")
```

You should now see a pop-up window that lets you customize your new workspace. You will need to give your new workspace a unique name and assign it to a billing project. The "anvil-outreach" billing project is used here as an example, but you will not be able to assign it. You'll have to use one of your own billing projects. After filling out these two fields, click the "Quick Create Workspace" button to create a workspace without enabling sharing or additional security options.

> ```{r, echo=FALSE, fig.alt="You will need to give your new workspace a unique name and assign it to a billing project. The anvil-outreach billing project is used here as an example, but you don't have permission to use it. You’ll have to use one of your own. After filling out these two fields, click the Quick Create Workspace to create your workspace without enabling sharing or additional security options."}

Reviewer: Want to anonymize any of this/the following screenshots?

Author: I have done so!

```{r, fig.align='center', echo = FALSE, fig.alt= "Choose the name of the workspace that you created earlier. Notice that the workspace name starts with the billing project. In our example, the workspace is 'anvil-outreach/import_data_sra'."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3ce9c5d5921_0_18#slide=id.g3ce9c5d5921_0_18")
```

After you have imported the workflow, you will see it as an option on the Workflow tab.

```{r, fig.align='center', echo = FALSE, fig.alt= "After you have imported the workflow, you will see it as an option on the Workflow tab."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3ce9c5d5921_0_94#slide=id.g3ce9c5d5921_0_94")
```

## Step Three: Import sample TSV file

The WDL you have chosen requires a list of the SRA samples you want to import into your workspace. You can do this sample by sample, but a more efficient way is to save the sample IDs to a TSV file and let AnVIL run parallel workflows for each sample. 

> 

Reviewer: ```suggestion

::: {.notice}
You can customize which samples you want for your workflow later.
:::

```

Let's take a moment to get acquainted with the new file we'll be uploading.

First, download the samples file here. You might have to right-click and "Save as".

> TODO: Add link to downloadable file

Reviewer: Did you mean to add this?

Next, open the file on your local machine. This is what it might look like in Microsoft Excel:

```{r, fig.align='center', echo = FALSE, fig.alt= "The samples we want to import from SRA are listed in rows in `samples.tsv, Make sure the column header is 'entity:sample_id'."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3ce9c5d5921_0_24#slide=id.g3ce9c5d5921_0_24")
```

::: {.notice}

> The column header `entity:sample_id` is important. `entity:` is required. `samples` becomes the name of the data table. So for example, if our header was `entity:reference_id`, a data table called "reference" would be created in AnVIL. If you didn't want to overwrite anything in the original "samples" table, you could change the column header. As long as none of the IDs are the same, no data will be overwritten. 

Reviewer: ```suggestion
The column header `entity:sample_id` is important. `entity:` is required. `sample` becomes the name of the data table. So for example, if our header was `entity:reference_id`, a data table called "reference" would be created in AnVIL. If you didn't want to overwrite anything in the original "samples" table, you could change the column header. As long as none of the IDs are the same, no data will be overwritten. 
```

ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3ce9c5d5921_0_27#slide=id.g3ce9c5d5921_0_27")
```

This opens a popup that looks like this:

```{r, fig.align='center', echo = FALSE, fig.alt= "The popup is titled Import Data Table and has the option to click to select a .tsv file"}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3ce9c5d5921_0_30#slide=id.g3ce9c5d5921_0_30")
```

Click to select your samples.tsv file. You will see a warning about potentially overwriting the existing entries. We know that none of the IDs in the new samples file overlap, so click START IMPORT JOB.

Now when we look at the TABLE in the Data tab, we can see that our samples have been added!

```{r, fig.align='center', echo = FALSE, fig.alt= "After we have imported the tsv file, the samples table in our workspace will be populated."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3ce9c5d5921_0_33#slide=id.g3ce9c5d5921_0_33")
```

> ## Step Four: Launch workflow

Reviewer: One thought could be rearranging step 2 and 3 so we only go to the workflows tab once...

Author: Part of me likes that idea, but the other part of me likes having them get the workflows first because it reinforces the idea that importing a workflow to the workspace means the workflow is always there and can be accessed separately (instead of the idea that we have to import the workflow each time we want to add samples to the workspace).

We are finally ready to launch the workflow! Navigate to the Workflows tab and choose the WDL we imported previously.

```{r, fig.align='center', echo = FALSE, fig.alt= "On the Workflows tab, click on the WDL we imported earlier."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3ce9c5d5921_0_71#slide=id.g3ce9c5d5921_0_71")
```

> Most of the default options are fine for now, but we do want to tell the workflow to pull the sample names from the table we created earlier. Choose the radio button to select "Run workflow(s) with inputs defined by data table". You will be prompted to choose either "sample" or "sample_set" from the pull down menu. Either option will work, but for ease we will choose "sample". After making these changes, click the "Select Data" button under Step 2.

Reviewer: I wouldn't even mention this, just click on the "Select Data" box

Author: Change made

ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3ce9c5d5921_0_74#slide=id.g3ce9c5d5921_0_74")
```

After you have chosen the samples, the sample input field for the workflow will now read "this.sample_id". This just means that this workflow will run for each ID in the data table. The other input fields can be left as "optional". Click Save.

```{r, fig.align='center', echo = FALSE, fig.alt= "The sample input field for the workflow will be automatically populated after the samples have been chosen."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3ce9c5d5921_0_77#slide=id.g3ce9c5d5921_0_77")
```

Once the inputs have been saved, you can start the workflow by clicking the Launch button.

```{r, fig.align='center', echo = FALSE, fig.alt= "Click the Launch button to start the workflow."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3ce9c5d5921_0_80#slide=id.g3ce9c5d5921_0_80")
```

> A pop-up window will be appear to verify that you want to launch the workflow. This window will also include details about any cost thresholds you did or did not set on your workflow. (We did not set any, but some workflows will incur charges. The box for setting a cost threshold is located just underneath the choose sample option on the workflow setup page.) Click Launch.

Reviewer: ```suggestion
A pop-up window will be appear to verify that you want to launch the workflow. This window will also include details about any cost thresholds you did or did not set on your workflow. (We did not set any, but it's important to remember that workflows will incur charges. The box for setting a cost threshold is located just underneath the choose sample option on the workflow setup page.) 

Click Launch.
```

Author: Added

ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3ce9c5d5921_0_118#slide=id.g3ce9c5d5921_0_118")
```

The results of each individual run will be stored within a subfolder of the workflow submission name.

```{r, fig.align='center', echo = FALSE, fig.alt= "The results of each individual run will be stored within a subfolder of the workflow submission name."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3ce9c5d5921_0_112#slide=id.g3ce9c5d5921_0_112")
```

You may have to click through several levels of directories to find the actual datafile. In this case, the results are within a folder labeled "shard-0".

```{r, fig.align='center', echo = FALSE, fig.alt= "You may have to click through several levels of directories to find the actual datafile. In this case, the results are within a folder labeled 'shard-0'."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3ce9c5d5921_0_115#slide=id.g3ce9c5d5921_0_115")
```

> When you click on the "pull_results.txt" file, you have the option to download the FASTQ files the workflow retrieved from SRA. Downloading the files from AnVIL (also called "data egress") will cost a small amount. However, you do not have to download them in order to work with the data within an AnVIL computing environment.

Reviewer: Did these actually work? This file is only a few bytes
